client/matching/isolation_loadbalancer.go

```diff
@@ -0,0 +1,145 @@
+	fallback           LoadBalancer
+	allIsolationGroups func() []string
+}
+
+func NewIsolationLoadBalancer(fallback LoadBalancer, provider PartitionConfigProvider, allIsolationGroups func() []string) LoadBalancer {
+	return &isolationLoadBalancer{
+		provider:           provider,
+		fallback:           fallback,
+		allIsolationGroups: allIsolationGroups,
+	}
+}
+
+func (i *isolationLoadBalancer) PickWritePartition(taskListType int, req WriteRequest) string {
+	taskList := *req.GetTaskList()
+	nPartitions := i.provider.GetNumberOfWritePartitions(req.GetDomainUUID(), taskList, taskListType)
+	taskListName := req.GetTaskList().Name
+
+	if nPartitions <= 1 {
+		return taskListName
+	}
+
+	taskGroup, ok := req.GetPartitionConfig()[partition.IsolationGroupKey]
+	if !ok {
+		return i.fallback.PickWritePartition(taskListType, req)
+	}
+
+	partitions, ok := i.getPartitionsForGroup(taskGroup, nPartitions)
+	if !ok {
+		return i.fallback.PickWritePartition(taskListType, req)
+	}
+
+	p := i.pickBetween(partitions)
+
+	return getPartitionTaskListName(taskList.GetName(), p)
+}
+
+func (i *isolationLoadBalancer) PickReadPartition(taskListType int, req ReadRequest, isolationGroup string) string {
+	taskList := *req.GetTaskList()
+	nRead := i.provider.GetNumberOfReadPartitions(req.GetDomainUUID(), taskList, taskListType)
+	taskListName := taskList.Name
+
+	if nRead <= 1 {
+		return taskListName
+	}
+
+	partitions, ok := i.getPartitionsForGroup(isolationGroup, nRead)
+	if !ok {
+		return i.fallback.PickReadPartition(taskListType, req, isolationGroup)
+	}
+
+	// Scaling down, we need to consider both sets of partitions
+	if numWrite := i.provider.GetNumberOfWritePartitions(req.GetDomainUUID(), taskList, taskListType); numWrite != nRead {
+		writePartitions, ok := i.getPartitionsForGroup(isolationGroup, numWrite)
+		if ok {
+			for p := range writePartitions {
+				partitions[p] = struct{}{}
+			}
+		}
+	}
+
+	p := i.pickBetween(partitions)
+
+	return getPartitionTaskListName(taskList.GetName(), p)
+}
+
+func (i *isolationLoadBalancer) UpdateWeight(taskListType int, req ReadRequest, partition string, info *types.LoadBalancerHints) {
+}
+
+func (i *isolationLoadBalancer) getPartitionsForGroup(taskGroup string, partitionCount int) (map[int]any, bool) {
+	if taskGroup == "" {
+		return nil, false
+	}
+	isolationGroups := slices.Clone(i.allIsolationGroups())
+	slices.Sort(isolationGroups)
+	index := slices.Index(isolationGroups, taskGroup)
+	if index == -1 {
+		return nil, false
+	}
+	partitions := make(map[int]any, 1)
+	// 3 groups [a, b, c] and 4 partitions gives us a mapping like this:
+	// 0: a
+	// 1: b
+	// 2: c
+	// 3: a
+	if len(isolationGroups) <= partitionCount {
+		for j := index; j < partitionCount; j += len(isolationGroups) {
+			partitions[j] = struct{}{}
+		}
+		// 4 groups [a,b,c,d] and 3 partitions gives us a mapping like this:
+		// 0: a, d
+		// 1: b
+		// 2: c
+	} else {
+		partitions[index%partitionCount] = struct{}{}
+	}
+	return partitions, true
```

Comment:
I know the above code will always populate `partitions`. However in case that logic evolves and it becomes possible that `partitions` is empty here we should do size check and return false if empty